File: traceharbor-hooks/src/main/java/io/traceharbor/hook/art/RuntimeVerifyMute.java
package io.traceharbor.hook.art;

import androidx.annotation.Nullable;

import io.traceharbor.hook.HookManager.NativeLibraryLoader;
import io.traceharbor.util.TraceHarborLog;

/**
 * Created by tomystang on 2022/11/16.
 */
public final class RuntimeVerifyMute {
    private static final String TAG = "TraceHarbor.RuntimeVerifyMute";

    public static final RuntimeVerifyMute INSTANCE = new RuntimeVerifyMute();

    private NativeLibraryLoader mNativeLibLoader = null;
    private boolean mNativeLibLoaded = false;

    public RuntimeVerifyMute setNativeLibraryLoader(@Nullable NativeLibraryLoader loader) {
        mNativeLibLoader = loader;
        return this;
    }

    private boolean ensureNativeLibLoaded() {
        synchronized (this) {
            if (mNativeLibLoaded) {
                return true;
            }
            try {
                if (mNativeLibLoader != null) {
                    mNativeLibLoader.loadLibrary("traceharbor-hookcommon");
                    mNativeLibLoader.loadLibrary("traceharbor-artmisc");
                } else {
                    System.loadLibrary("traceharbor-hookcommon");
                    System.loadLibrary("traceharbor-artmisc");
                }
                mNativeLibLoaded = true;
            } catch (Throwable thr) {
                TraceHarborLog.printErrStackTrace(TAG, thr, "Fail to load native library.");
                mNativeLibLoaded = false;
            }
            return mNativeLibLoaded;
        }
    }

    public boolean install() {
        if (!ensureNativeLibLoaded()) {
            return false;
        }
        return nativeInstall();
    }

    private static native boolean nativeInstall();

    private RuntimeVerifyMute() { }
}

File: traceharbor-hooks/src/main/java/com/kernelflux/traceharbor/hook/art/RuntimeVerifyMute.kt
package com.kernelflux.traceharbor.hook.art

import com.kernelflux.traceharbor.hook.HookManager.NativeLibraryLoader
import com.kernelflux.traceharbor.util.TraceHarborLog

/**
 * Created by tomystang on 2022/11/16.
 */
object RuntimeVerifyMute {
    private const val TAG = "TraceHarbor.RuntimeVerifyMute"

    private var nativeLibLoader: NativeLibraryLoader? = null
    private var nativeLibLoaded = false

    fun setNativeLibraryLoader(loader: NativeLibraryLoader?): RuntimeVerifyMute {
        nativeLibLoader = loader
        return this
    }

    private fun ensureNativeLibLoaded(): Boolean {
        synchronized(this) {
            if (nativeLibLoaded) {
                return true
            }
            try {
                if (nativeLibLoader != null) {
                    nativeLibLoader?.apply {
                        loadLibrary("traceharbor-hookcommon")
                        loadLibrary("traceharbor-artmisc")
                    }
                } else {
                    System.loadLibrary("traceharbor-hookcommon")
                    System.loadLibrary("traceharbor-artmisc")
                }
                nativeLibLoaded = true
            } catch (thr: Throwable) {
                TraceHarborLog.printErrStackTrace(TAG, thr, "Fail to load native library.")
                nativeLibLoaded = false
            }
            return nativeLibLoaded
        }
    }

    fun install(): Boolean {
        if (!ensureNativeLibLoaded()) {
            return false
        }
        return nativeInstall()
    }

    private external fun nativeInstall(): Boolean
}

